File: rho.py
from fractions import gcd
import random

random.seed(17)

def pollard_rho(n):
    i = 0
    while True:
        c = random.randint(2, n-1)
        f = lambda x: x**2 + c 
        x = y = 2 
        d = 1 
        while d == 1:
            x = f(x) % n 
            y = f(f(y)) % n 
            d = gcd(abs(x - y), n)
            i += 1
        if d != n: return d, i

factor, cnt = pollard_rho(int(input()))
print("Factor {} found in {} iterations.".format(factor, cnt))

File: rho.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"github.com/lionell/pollard/algo"
	"math/big"
	"os"
)

var concurrencyLimit = flag.Int("j", 1, "max number of sub-routines to run concurrently")

func main() {
	flag.Parse()

	// read input
	s := bufio.NewScanner(os.Stdin)
	s.Scan()

	// try parse number
	n, ok := new(big.Int).SetString(s.Text(), 10)
	if !ok {
		fmt.Printf("Error while parsing input number %v", s.Text())
		return
	}

	// Check for primality. 10 iterations of Miller-Rabin + Baillie-PSW
	// is enough to say with certainty 1 - (1/4)^10
	if n.ProbablyPrime(10) {
		fmt.Printf("%v is prime.\n", n)
		return
	}

	// finally, try to factorize
	f := algo.FactorizeParallel(n, *concurrencyLimit)
	fmt.Printf("Factor %v found.\n", f)
}
